src/fft.rs

```diff
@@ -0,0 +1,299 @@
+        /// ```
+        /// # References
+        ///
+        /// This implementation is based on the concepts discussed in
+        /// [Levente Kovács' post](https://kovleventer.com/blog/fft_real/).
+        pub fn $func_name(
+            input_re: &[$precision],
+            output_re: &mut [$precision],
+            output_im: &mut [$precision],
+        ) {
+            assert!(output_re.len() == output_im.len() && input_re.len() == output_re.len());
+            let big_n = input_re.len();
+
+            // Splitting odd and even
+            let (mut z_even, mut z_odd): (Vec<_>, Vec<_>) =
+                input_re.chunks_exact(2).map(|c| (c[0], c[1])).unzip();
+
+            // Z = np.fft.fft(z)
+            $fft_func(&mut z_even, &mut z_odd, Direction::Forward);
+
+            let mut z_x_re = Vec::with_capacity(z_even.len());
+            let mut z_x_im = Vec::with_capacity(z_even.len());
+            let mut z_y_re = Vec::with_capacity(z_even.len());
+            let mut z_y_im = Vec::with_capacity(z_even.len());
+
+            z_x_re.push(0.0);
+            z_x_im.push(0.0);
+            z_y_re.push(0.0);
+            z_y_im.push(0.0);
+
+            // Zminconj = np.roll(np.flip(Z), 1).conj()
+            // Zx =  0.5  * (Z + Zminconj)
+            // Zy = -0.5j * (Z - Zminconj)
+            z_even
+                .iter()
+                .skip(1)
+                .zip(z_odd.iter().skip(1))
+                .zip(z_even.iter().skip(1).rev())
+                .zip(z_odd.iter().skip(1).rev())
+                .for_each(|(((z_e, z_o), z_e_mc), z_o_mc)| {
+                    let a = *z_e;
+                    let b = *z_o;
+                    let c = *z_e_mc;
+                    let d = -(*z_o_mc);
+
+                    let t = 0.5 * (a + c);
+                    let u = 0.5 * (b + d);
+                    let v = -0.5 * (a - c);
+                    let w = 0.5 * (b - d);
+
+                    z_x_re.push(t);
+                    z_x_im.push(u);
+                    z_y_re.push(w);
+                    z_y_im.push(v);
+                });
+
+            let a = z_even[0];
+            let b = z_odd[0];
+            let c = z_even[0];
+            let d = -z_odd[0];
+
+            let t = 0.5 * (a + c);
+            let u = 0.5 * (b + d);
+            let v = -0.5 * (a - c);
+            let w = 0.5 * (b - d);
+
+            z_x_re[0] = t;
+            z_x_im[0] = u;
+            z_y_re[0] = w;
+            z_y_im[0] = v;
+
+            let (twiddle_re, twiddle_im): (Vec<$precision>, Vec<$precision>) =
+                generate_twiddles(big_n / 2, Direction::Forward);
+
+            // Zall = np.concatenate([Zx + W*Zy, Zx - W*Zy])
+
+            for i in 0..big_n / 2 {
+                let zx_re = z_x_re[i];
+                let zx_im = z_x_im[i];
+                let zy_re = z_y_re[i];
+                let zy_im = z_y_im[i];
+                let w_re = twiddle_re[i];
+                let w_im = twiddle_im[i];
```

Comment:
Indexing in a potentially hot loop need to be inspected for bounds checks in godbolt or similar, and a comment added that it's inspected and doesn't have them, or that a `.get_unchecked()` prototype didn't gain you any performance.

If bounds checks do get in the way, try [the usual techniques](https://shnatsel.medium.com/how-to-avoid-bounds-checks-in-rust-without-unsafe-f65e618b4c1e) or amortize them with processing batches of 4 or 8 elements at a time, performing the bounds checks once per slice.